### PR TITLE
feat: Added DialContext to Client and proxy package

### LINKED ISSUE #431: Add Proxy.DialContext(..) method
## Feature Description

Right now `proxy.Proxy` exposes a `func Dial(instance string) (net.Conn, error)` that is registered using [`mysql.RegisterDial(...)`](https://github.com/go-sql-driver/mysql/blob/46351a8892976898935f653f5333782579a96fa5/driver.go#L63) which is [deprecated](https://github.com/go-sql-driver/mysql/blob/46351a8892976898935f653f5333782579a96fa5/driver.go#L62) in favor of [`mysql.RegisterDialContext(...)`](https://github.com/go-sql-driver/mysql/blob/46351a8892976898935f653f5333782579a96fa5/driver.go#L49).

A context-unaware dialling in production has caused us uncontrollably long requests during the failover process of MySQL instances.

A clear and concise description of what feature you would like to see, and why it would be useful to have added.

## Alternatives Considered

We though about instantiating a `*proxy.Proxy{}` in each dial, providing as `Dialer` a `(*net.Dialer{}).DialContext(...)` function curried with the provided context, but since `proxy.Proxy` holds a cache of configs, that would lead to a config retrieval in each of the dials.

## Additional Context

I would be happy to help with this. 

As I can see, we could add a `ContextDialer func(ctx context.Context, net, addr string) (net.Conn, error)` field to `Proxy`, and keep falling back to `Dialer` if this one isn't set. We'd expose a new receiver `DialContext(ctx context.Context, instance string) (net.Conn, error)` on the `Proxy` and the same function at package-level, making the previous `Dial` functions call this one with `context.Background()`. This new functions would be the ones registered as default MySQL dialers for CloudSQL.

In order to support the context in the config retrieval mechanism, we can expose a `RemoteContext` function alternative to `Remote` on the `RemoteCertSource` and as adding that new method to the `CertSource` interface would be a breaking change (which we don't want, I guess) we can check in runtime whether the cert source provided implements the `interface { RemoteContext(...) }` and call it in this case, properly documenting that feature on the `CertSource` interface.

### PR DESCRIPTION
**Pull request by colega**
## Change Description

Adds DialContext method to client, using the optional ContextDialer.
Also checks whether `proxy.FromEnvironment()` returns a ContextDialer,
using it in that case.

Changed mysql hook to register the `proxy.DialContext` as dialer.

## Checklist

- [x] Make sure to open an issue as a 
  [bug/issue](https://github.com/GoogleCloudPlatform/cloudsql-proxy/issues/new/choose) 
  before writing your code!  That way we can discuss the change, evaluate 
  designs, and agree on the general idea.
- [x] Ensure the tests and linter pass
- [x] Appropriate documentation is updated (if necessary)

## Relevant issues:

- Fixes #431 

## Notes

I didn't go deep into adding context-versions of all the functions provided to avoid over-polluting the API, the ones I skipped were:
- `mysql.Dial`, `DialPassword`, `DialCfg`: could be `DialContext`, `DialContextPassword`, etc, however since the configs they use for `DialCfg` don't set the `Timeout` that would only be useful for the `db.Ping()` call, and IMO, whoever is interested in timeouts would build their own `mysql.Config` and call `open.DB`.  
- `proxy.Init(auth *http.Client, connset *ConnSet, dialer Dialer)` could have an `InitContextDialer` accepting a `ContextDialer` instead of `Dialer`, however, the same can be achieved with `InitWithClient` so I just mentioned that in a comment.

If some of those are still interesting, I can implement them ☝️ 

I didn't mark the non-contextual versions as _deprecated_ since I saw that they're the only way of dialing for postgres so it would be weird to have them deprecated.

Finally, I modified the comments on the non-context-aware versions of functions referencing them to the context versions, to avoid duplicating the documentation.

### CODE REVIEW

File: proxy/proxy/client_test.go
Comment (kurtisvg) [inline proxy/proxy/client_test.go:86]: Not sure I follow this logic. Can you elaborate what's going on here?

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: proxy/proxy/client.go
Comment (colega) [reply proxy/proxy/client.go]: I just extracted it to make `tryConnect` (opinionatedly) cleaner.

Moved the logic back there and removed `c.dialer()` at all. 👍

File: proxy/proxy/client_test.go
Comment (colega) [reply proxy/proxy/client_test.go]: If you're asking then it obviously needed a rework. I was passing a canceled context and checking the error to be `context.Canceled` since that's what our dialer was returning (`ctx.Err()`) although incorrectly, since `proxy.FromEnvironment()` would return that error too.

Changed it to return `errFakeDial` instead just like the `Dialer` test does.

**Comment by colega** [comment]
@kurtisvg applied the requested changed, PTAL.

**Review by kurtisvg — APPROVED** [review]

File: proxy/proxy/client.go
Comment (Carrotman42) [inline proxy/proxy/client.go:302]: We should consider adding a "Deprecated" notice here, since people should ideally be plumbing contexts correctly these days.

Same thought for adding a deprecated message to all of the existing things which are now duplicated into context and non-context version. If you ask me, people who still aren't plumbing contexts (especially for i/o) are just asking for tricky bugs down the line :)  (which is why I think this PR is great, thanks again for sending it out.)

File: proxy/proxy/dial.go
Comment (Carrotman42) [inline proxy/proxy/dial.go:69]: Can the code just use the x/net/proxy.ContextDialer that already exists? I think that's whats happening, I can't tell where this type is being used actually.

File: proxy/proxy/client.go
Comment (Carrotman42) [reply proxy/proxy/client.go]: FWIW I disagree with Kurtis here. It's his call ultimately, and since the code is currently like this feel free to leave it, but: this tryConnect function is quite complicated in part because there's no clear separation of concerns: it's currently in charge of a lot of things. I think separating the code out into a separate function which is in charge of obtaining the dialer is a better decision (after this change, not before) because this operation now has a lot of branching and decisions - even if it's just used once, it's good to organize complicated code so that there's a clear API that future maintainers can wrap their head around.

Also, I think the original name was good: it's quite common to have a field named X and a method named x which accesses that field and does something special with it (and/or other fields). The similar naming is an intentional nod to the fact that, yes, the `dialer` method probably accesses the `Dialer` field. I'd only find it confusing if the `dialer` method didn't access the `Dialer` field, I guess.  But any other name works too, I don't really care what it's called as long as it's not part of the exported API.

(Also, I can't see the original change because it was force-pushed, so I'm just guessing what it looked like; please prefer to not force-push, as it makes code reviews harder to do because it's harder to see the full history after the fact.)

Comment (kurtisvg) [reply proxy/proxy/client.go]: So the root of my concern here is that the public API for this package is a bit less well defined (or at least understood by myself) and I'm not entirely sure if use of the `Client` struct is reasonable accessed by any downstream dependencies outside of this package. 

With that in mind, it seems confusing to me to have a public (exported) `*Client.Dialer` and `*Client.DialerContext` attributes, but then have a private (non-exported) `c *Client.dialer() ContextDialer` determining which are used. I wanted to avoid this function being exported at some point in the future with the context of the conversation lost and leading to a confusing or possible changed API surface. 

So my objection is less about where the code is structurally organized, and more of trying to keep the overall namespace a little more clear.

Comment (colega) [reply proxy/proxy/client.go]: Hey, the original commit can still be seen in the PR timeline in github: https://github.com/GoogleCloudPlatform/cloudsql-proxy/commit/68542d0e4a6fc6b98a73a5741169cd3bd82d614e

Wasn't sure about force pushing, or adding a `refactor:` commit (refactoring something I'm proposing? Then I'm forcing to squash) but yeah, I see your point.

File: proxy/proxy/client.go
Comment (colega) [reply proxy/proxy/client.go:302]: I considered adding that, but since this repo is using this for Postgres driver I thought it would be weird to say this is deprecated.

I can add the message if you want you to.

File: proxy/proxy/client.go
Comment (colega) [reply proxy/proxy/client.go]: We can always `git checkout 68542d0 - proxy/proxy/client.go && git commit proxy/proxy/client.go -m "refactor: Extract dialer selection logic into Client.dialer()` :)

File: proxy/proxy/dial.go
Comment (colega) [reply proxy/proxy/dial.go]: Oh right this code is completely unused. Too much copy-paste I guess. Removed in https://github.com/GoogleCloudPlatform/cloudsql-proxy/pull/483/commits/eec491fd6b9296a6f3d8d160bb16ab44a0e82a45

File: proxy/proxy/client.go
Comment (colega) [reply proxy/proxy/client.go]: I've extract back the logic in https://github.com/GoogleCloudPlatform/cloudsql-proxy/pull/483/commits/393b14a9bf4fa195de47ca164a8844ddd852219c in order to keep `tryConnect` clean, and I renamed the function into `selectDialer()` to address @kurtisvg's concerns about colliding naming.

Comment (Carrotman42) [reply proxy/proxy/client.go]: > I wanted to avoid this function being exported at some point in the future with the context of the conversation lost and leading to a confusing or possible changed API surface.

Note that it would be a compile error to have a Dialer method on a struct with a Dialer field; so I think this is not actually a problem.

selectDialer seems like an OK name too.

Comment (Carrotman42) [reply proxy/proxy/client.go]: > Hey, the original commit can still be seen in the PR timeline in github: 68542d0

Ah thanks, I didn't think to look there. I clicked on the conversation and github claimed that the code it was on didn't exist, so I trusted the web UI :)

As for squashing, all the open source repositories I'm involved with prefer to squash-and-commit at the end, so that it doesn't matter how many commits it took during review. This keeps the main tree's history clean (one commit per PR), while still allowing individual commits to be seen in the PR itself if anyone is interested in the history.

**Review by kurtisvg — APPROVED** [review]

File: proxy/proxy/client.go
Comment (kurtisvg) [reply proxy/proxy/client.go]: I prefer an intact history via separate commits, but greatly prefer use rebasing over merge when updating upstream - but that particular bikeshed doesn't need to be painted in this thread :)

File: proxy/proxy/client.go
Comment (kurtisvg) [reply proxy/proxy/client.go:302]: I'm not particularly concerned about this. We can add it in the future if it becomes a serious issue.

**Comment by kurtisvg** [comment]
Thanks again for your work on this @colega!

**Pull request merged**
